sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -176,7 +176,10 @@ case class AdaptiveSparkPlanExec(
                 }
               }(AdaptiveSparkPlanExec.executionContext)
             } catch {
-              case e: Throwable => events.offer(StageFailure(stage, e))
+              case e: Throwable =>
+                errors.prepend(
+                  new SparkException(s"Early failed query stage found: ${stage.treeString}", e))
+                cleanUpAndThrowException(errors, Some(stage.id))
```

Comment:
nit (and apologize for my previous mistake): at this point, we won't have any other errors, lets simplify it to:
```
val ex = new SparkException(s"Early failed query stage found: ${stage.treeString}", e))
cleanUpAndThrowException(Seq(ex), Some(stage.id))
```

Reply:
But `errors` may come from previous materialized but failed stages?

Reply:
You call `cleanUpAndThrowException` right after, so any error in this round has not been added through "StageFailure" message processing yet. And there cannot be any errors from previous rounds, otherwise it would have thrown and exited already.

Reply:
Make sense!